cpp/include/raft/neighbors/ivf_pq.cuh

```diff
@@ -117,6 +145,39 @@ inline auto extend(raft::device_resources const& handle,
     handle, orig_index, new_vectors, new_indices, n_rows);
 }
 
+/**
+ * @brief Build a new index containing the data of the original plus new extra vectors.
+ *
+ * Implementation note:
+ *    The new data is clustered according to existing kmeans clusters, then the cluster
+ *    centers are unchanged.
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices in the source dataset
+ *
+ * @param handle
+ * @param orig_index original index
+ * @param[in] new_vectors a device matrix view to a row-major matrix [n_rows, index.dim()]
+ * @param[in] new_indices a device matrix view to a vector of indices [n_rows].
+ *    If the original index is empty (`orig_index.size() == 0`), you can pass `nullptr`
+ *    here to imply a continuous range `[0...n_rows)`.
+ *
+ * @return the constructed extended ivf-pq index
+ */
+template <typename T, typename IdxT>
+inline auto extend(raft::device_resources const& handle,
+                   const index<IdxT>& orig_index,
+                   const raft::device_matrix_view<const T, IdxT, row_major>& new_vectors,
+                   const raft::device_matrix_view<const IdxT, IdxT, row_major>& new_indices)
+  -> index<IdxT>
+{
+  IdxT n_rows = new_vectors.extent(0);
+  ASSERT(n_rows == new_indices.extent(0),
```

Comment:
Maybe we could also check here that the number of columns in `new_vectors` matches the number of dimensions in the index?